README.md
python-utils
============

collection of usefull utilities written by me in python 

### module_iec_xml_generator

Xml generation using lxml
designed for Module MEK by energosoyuz
input data: csv file

### cts_complex_sql_get

This program get data from MSSQL database by given
parameters and calculates avarage.

Target is MSSQL 2005 database of complex from
http://www.cts.spb.ru/

output: 2 files with calculated values

### mssql_zero_fill

Fill MSSQL 2005 database with zeros by given parameters


tangotango-theme-create/changes
"black" - "#000000"
"brown" - "#a52a2a"
"brown4" - "#8b2323"
"Coral" - "#ff7f50"
"cyan" - "#00ffff"
"dark grey" - "#a9a9a9"
"dark red" - "#8b0000"
"dark slate blue" - "#555753"
"DimGray" - "#696969"
"dodger blue" - "#1e90ff"
"DodgerBlue3" - "#1874cd"
"ForestGreen" - "#228b22"
"Gold" - "#ffd700"
"gray30" - "#4d4d4d"
"green" - "#00ff00"
"grey" - "#bebebe"
"grey10" - "#1a1a1a"
"grey15" - "#262626"
"grey20" - "#333333"
"grey40" - "#666666"
"grey50" - "#7f7f7f"
"grey60" - "#999999"
"grey70" - "#b3b3b3"
"grey90" - "#e5e5e5"
"indian red" - "#cd5c5c"
"IndianRed" - "#cd5c5c"
"khaki4" - "#8b864e"
"light cyan" - "#e0ffff"
"LightSkyBlue3" - "#009acd"
"lightyellow" - "#ffffe0"
"lime green" - "#32cd32"
"magenta" - "#ff00ff"
"magenta3" - "#cd00cd"
"magenta4" - "#8b008b"
"orange" - "#ffa500"
"PaleTurquoise4" - "#668b8b"
"red" - "#ff0000"
"Red" - "#ff0000"
"red1" - "#ff0000"
"red4" - "#8b0000"
"RosyBrown1" - "#ffc1c1"
"rotal blue" - "#4169e1"
"royal blue" - "#4169e1"
"sienna4" - "#8b4726"
"skyblue2" - "#7ec0ee"
"SkyBlue4" - "#00688b"
"tomato" - "#ff6347"
"turquoise" - "#40e0d0"
"violet" - "#ee82ee"
"white" - "#ffffff"
"White" - "#ffffff"
"yellow" - "#ffff00"
"yellow1" - "#ffff00"
"DejaVu Sans Mono" - "DejaVu Sans Mono"
"unknown" - "unknown"
"Sans Serif" - "Sans Serif"
"#fce94f" - "#ffaf00"
"#483d8b" - "#555753"
"#4a708b" - "#00688b"
"#edd400" - "#fcaf3d"
"#afeeee" - "#668b8b"
"#eeeeec" - "#d3d7cf"


mssql_zero_fill/mssql_zero_fill.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: mssql_zero_fill.py 
# Author: Yegor Bayev (kodx) <kodxxx@gmail.com> 
# Created: Wed Dec 19 16:21:48 2012
# License: 
# Copyright 2012 Yegor Bayev (kodx) <kodxxx@gmail.com>
#      
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#     
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#      
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.

"""
Fill MSSQL 2005 database with zeros by given parameters
"""

import _mssql
import time
from datetime import datetime, timedelta

sqluser = 'user'
sqlpass = 'password'
sqlsrv = '192.168.0.1'
sqldb = 'dbname'
sqlenc = 'cp1251'

def mkdate(text):
    return datetime.strptime(text, '%Y-%m-%d %H:%M:%S')

def insert_sql(dDate, dItem):
    sqlquery = """
    insert into data
    (parnumber, object, item, value0, value1, objtype, data_date,
    rcvstamp, season, p2kstatus, p2kstatush, appid)
    values
    (12, 4372, %(dItem)d, 0.0, 0.0, 0, '%(sDate)s', '%(sRcvDate)s',
    4023, 0, 0, 38) 
    """ % \
    {'sDate': dDate.strftime('%Y-%m-%d %H:%M:%S'),
     'sRcvDate': (datetime.now()).strftime('%Y-%m-%d %H:%M:%S'),
     'dItem': dItem}
    try:
        conn = _mssql.connect(server=sqlsrv, user=sqluser, password=sqlpass, \
                              database=sqldb, charset=sqlenc)
        conn.execute_non_query(sqlquery)
        if conn.identity == None:
            print "something is wrong, insert", sqlquery, " failed"
            

    except _mssql.MssqlDatabaseException,e:
        print 'connection error number=', e, ', severity=', e.severity
    finally:
        conn.close()

    return 0
    
if __name__ == '__main__':
    #curDt = mkdate('2012-12-17 01:00:00')
    curDt = datetime.now()
    dtCur = (curDt - timedelta(days=1)).replace(hour=0, minute=0,
                                                  second=0,
                                                  microsecond=0)
    for i in xrange(48):
        dtCur += timedelta(minutes=30)
        insert_sql(dtCur, 1)
        insert_sql(dtCur, 3)

# mssql_zero_fill.py ends here



cts_complex_sql_get/cts_complex_sql_get.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# cts_complex_sql_get.py
#      
# Copyright 2012 Yegor Bayev (kodx) <kodxxx@gmail.com>
#      
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#     
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#      
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.

import _mssql
import time
from datetime import datetime, timedelta
import csv

"""
This program get data from MSSQL database by given
parameters and calculates avarage.

Target is MSSQL 2005 database of complex from
http://www.cts.spb.ru/

output: create 2 files:
full - all data from request
avg - average

it uses pymssql module based upon freetds and it can be broken for
database above version 2008

http://code.google.com/p/pymssql/
python 2.6
"""

# source date and time
sStart = '2013-01-16 07:00:00'
sEnd = '2013-01-16 18:00:00'

# Table to get data from
# SNAP_1 - every minute snapshot
# TIT - all data
sqltable = 'SNAP_1'

# specific parameter
uidvar = 291
uidvarDesc = 'G3' # this description will be added to filename
"""
Generator channels
G2 - 100
G3 - 291
G4 - 482
G5 - 673
G6 - 864
"""

# Database connection settings
sqluser = 'user'
sqlpass = 'password'
sqlsrv = 'mssql_base_address'
sqldb = 'dbname'
sqlenc = 'cp1251' # database encoding

# calc. of utc datetime offset, it can be set in numbers
# for this example datatime in db was in utc, and i had to convert it
# to local time (UTC+4)
utcOffset = (time.timezone / -(60*60))

# calculate date and time for request
dtStart = (datetime.strptime(sStart, '%Y-%m-%d %H:%M:%S') -
           timedelta(hours=utcOffset))
dtEnd = (datetime.strptime(sEnd, '%Y-%m-%d %H:%M:%S') -
         timedelta(hours=utcOffset))

# sql request
sqlquery = """
SELECT Tbl1.DateInKP AS DateINKP, Tbl1.ValueVariable 
FROM ( SELECT * FROM %(sqltable)s_%(dtStartShort)s
       where UIDVariable=%(uidvar)d
       UNION
       SELECT * FROM %(sqltable)s_%(dtEndShort)s
       where UIDVariable=%(uidvar)d
) Tbl1
where DateInKP >= '%(dtStart)s' and 
      DateInKP <  '%(dtEnd)s'
ORDER BY Tbl1.DateINKP""" % \
{'uidvar': uidvar,
 'sqltable': sqltable,
 'dtStartShort': dtStart.strftime('%Y%m%d'),
 'dtEndShort': dtEnd.strftime('%Y%m%d'),
 'dtStart': dtStart.strftime('%Y-%m-%d %H:%M:%S'),
 'dtEnd': dtEnd.strftime('%Y-%m-%d %H:%M:%S')}
# In this case i was using this form of string format cause standart
# connect() was broken.

def total_seconds(date):
    """
    Calculate seconds in given datetime.timedelta.
    Parameters:
    date - datetime.timedelta
    
    returns - float
    
    This function python 2.7.2 built-in, i was using python 2.6 and
    had to make my own implementation.
    """
    return (date.microseconds + (date.seconds + date.days * 24 * 3600)
            * 10**6) / 10**6.

def calc_avg_int(input_list):
    prev_val = count = tmpsum = 0
    tmp = input_list[0][0]
    res = []
    cur_hour = prev_time = datetime(tmp.year, tmp.month, tmp.day, tmp.hour)

    for row in input_list:
        if (row[0] - cur_hour) >= timedelta(hours=1):
            cur_hour = datetime(row[0].year, row[0].month, row[0].day,
                                row[0].hour)
            secs = total_seconds(cur_hour - prev_time)
            tmpsum += prev_val * secs
            res.append([cur_hour, tmpsum/3600])
            tmpsum = 0
        secs = total_seconds(row[0] - prev_time)
        tmpsum += float(row[1]) * secs
        prev_time = row[0]
        prev_val = float(row[1])
        
    tmp = prev_time + timedelta(hours=1)
    cur_hour = datetime(tmp.year, tmp.month, tmp.day, tmp.hour)
    tmpsum += prev_val * total_seconds(cur_hour - prev_time)
    res.append([cur_hour, tmpsum/3600])

    return res    

def calc_avg(input_list):
    count = tmpsum = 0
    tmp = input_list[0][0]
    res = []
    cur_hour = prev_time = datetime(tmp.year, tmp.month, tmp.day,
                                    tmp.hour)
    for row in input_list:
        if (row[0] - cur_hour) >= timedelta(hours=1):
            cur_hour = datetime(row[0].year, row[0].month, row[0].day,
                                row[0].hour)
            res.append([cur_hour, float(tmpsum/count)])
            tmpsum = count = 0
        tmpsum += float(row[1])
        count += 1
        prev_time = row[0]

    if (prev_time - cur_hour) < timedelta(hours=1):
        cur_hour = cur_hour+timedelta(hours=1)
    
    res.append([cur_hour, float(tmpsum/count)])
    return res

def read_sql():
    reslst = [] # result list
    try:
        conn = _mssql.connect(server=sqlsrv, user=sqluser, password=sqlpass, \
                              database=sqldb, charset=sqlenc)
        conn.execute_query(sqlquery)
        for row in conn:
            reslst.append([(row[0] + timedelta(hours=utcOffset)),
                           row[1]])

    except _mssql.MssqlDatabaseException,e:
        print 'connection error number=', e, ', severity=', e.severity
    finally:
        conn.close()
    
    return reslst

def mkcsv(fname, input_list):
    with open(fname, 'wb') as csvfile:
        csvwriter = csv.writer(csvfile, delimiter=';',
                                quotechar='|',
                                quoting=csv.QUOTE_MINIMAL)

        for row in input_list:
            csvwriter.writerow([row[0].strftime('%Y-%m-%d %H:%M:%S'),
                                str(row[1]).replace('.',',')])
        
if __name__ == '__main__':
    sdata = read_sql()
    if not sdata :
        print "No data to proceed, check SQL Query!"
    else:
        if sqltable == 'TIT':
            avgdata = calc_avg_int(sdata)
        elif sqltable == 'SNAP_1':
            avgdata = calc_avg(sdata)
        else:
            print 'Error table name selected!'
    
        stmp = sStart.replace(':', '_')
        etmp = sEnd.replace(':', '_')
        mkcsv('full '+uidvarDesc+' '+stmp+' - '+etmp+'.csv', sdata)
        mkcsv('avg '+uidvarDesc+' '+stmp+' - '+etmp+'.csv', avgdata)



module_iec_xml_generator/module_iec_xml_generator.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# module_iec_xml_generator.py
#      
# Copyright 2013 Yegor Bayev (kodx) <kodxxx@gmail.com>
#      
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#     
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#      
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.


"""
Simple xml generation program designed for Neva complex
input data - csv file.
http://www.energosoyuz.spb.ru/content/%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B0%D0%BC%D0%BC%D0%BD%D0%BE%D0%B5-%D0%BE%D0%B1%D0%B5%D1%81%D0%BF%D0%B5%D1%87%D0%B5%D0%BD%D0%B8%D0%B5
"""

try:
    from lxml import etree as et
except ImportError:
    print("can't find lxml package, please install it properly")  

if __name__ == '__main__':
    root = et.Element("IECSchema")

    iec = et.SubElement(root,
                        "IEC", Station="Neva", Port="2404",
                        ClientLimit="5", ASDUAddr="1", KP="1",
                        Sporadic="false", Period="1000")
    et.SubElement(iec, "Periodic").text = "true"
    
    srv = et.SubElement(root, "IECServers")
    et.SubElement(srv, "IECServer_id").text = "1"
    et.SubElement(srv, "IECServer_name").text = "Server"
    et.SubElement(srv, "IECServer_ip").text = "192.168.0.1"
    et.SubElement(srv, "IECServer_port").text = "2500"
    et.SubElement(srv, "IECServer_Period").text = "1"
    et.SubElement(srv, "IECServer_Mode")
    et.SubElement(srv, "IECServer_UniqueIDBase").text = "0"
    et.SubElement(srv, "IECServer_TimeMode").text = "1"
    et.SubElement(srv, "IECServer_NeedConvertToUTC").text = "0"
    et.SubElement(srv, "IECServer_PeriodSync").text = "60000"

    idx = 0
    for line in open("iec_values_and_opc_tags.csv", "rb"):
        values = (line.strip()).split(";")
        uid = values[0]
        tag = values[1]
        item = et.SubElement(root, "IECItems")
        et.SubElement(item, "IECItem_id").text = str(idx)
        et.SubElement(item, "IECItem_server").text = "1"
        et.SubElement(item, "IECItem_kp").text = "0"
        et.SubElement(item, "IECItem_uid").text = uid
        if int(uid) < 1000:
            itype = "TS"
        else:
            itype = "TI"
        
        et.SubElement(item, "IECItem_OPCName").text = itype+"."+tag
        et.SubElement(item, "IECItem_ordernum").text = "1"
        et.SubElement(item, "IECItem_UniqueID").text = uid
        et.SubElement(item, "IECItem_type").text = itype
        et.SubElement(item, "IECItem_Offset").text = "0"
        et.SubElement(item, "IECItem_Scale").text = "1"
        et.SubElement(item, "IECItem_UOM")
        idx += 1
    
    tree = et.ElementTree(root)
    tree.write("IEC_Config.xml", pretty_print=True, xml_declaration=True,
               encoding="UTF-8", standalone="true")
    
# module_iec_xml_generator ends here


tangotango-theme-create/tangotangokx-theme-create.py
#!/usr/bin/env python3
# coding: utf-8
# File: tangotangokx-theme-create.py
# Author: Yegor Bayev <baev.egor (at) gmail.com>
# Created: Thu Aug 29 20:44:21 2013
# License:
#      
# Copyright 2013 Yegor Bayev <baev.egor (at) gmail.com>
#      
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#     
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#      
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.

'''
create changes with command
$ cat ./tangotango-theme.el | grep -o '\"[(a-z)(A-Z)(0-9) ]*\"' | sort -u > changes
then fill corresponding colors to it number
'''

import fileinput, re

changesFile = 'changes'
sourceFile = 'tangotango-theme.el'
targetFile = 'tangotangokx-theme.el'

if __name__ == '__main__':
    chMap = {}
    with open(changesFile, 'r') as f:
        for line in f:
            if line.strip() != '':
                color, number = line.strip().split(' - ')
                chMap[color] = number
            
    p = re.compile('\"[(a-z)(A-Z)(0-9) #]*\"')
    
    with open(sourceFile, 'r') as f:
        outFile = []
        for line in f:
            m = p.findall(line.strip())
            if m:
                for i in m:
                    if i in chMap:
                        line = line.replace(i, chMap[i])

            # remove italic styles
            if ':slant italic' in line:
                line = line.replace(':slant italic', '')
            # remove font declaration
            if ':family "DejaVu Sans Mono" :foundry "unknown" :width normal :height 90 :weight normal ' in line:
                line = line.replace(':family "DejaVu Sans Mono" :foundry "unknown" :width normal :height 90 :weight normal ', '')
            # renaming theme
            if 'tangotango' in line:
                if 'color-theme-tangotango' not in line:
                    line = line.replace('tangotango', 'tangotangokx')
            # adding mod annotaion
            if ';; This file is NOT part of GNU Emacs.' in line:
                outFile.append(";; kodx version 0.0.2\n;; Author: Yegor Bayev <baev.egor (at) gmail.com>\n;; URL: https://github.com/kodx/emacs.d\n\n")
            outFile.append(line)

    with open(targetFile, 'w') as f:
        for line in outFile:
            f.write("%s" % line)
    
# tangotangokx-theme-create.py ends here
